### PR TITLE
Add internet detector utility

### PR DESCRIPTION
**Pull request by emtuls**
This adds a script to detect if internet connection is active or not by adding an icon in the taskbar and also changing the background based on if internet is detected or not.

This will also require an addition of a new python libraries to install:
`pywin32`

### CODE REVIEW

File: scripts/utils/internet_detect_tool.py
Comment (Ana06) [inline scripts/utils/internet_detect_tool.py:28]: We need to add this image to flare-vm and copy as we do with the background image.

File: scripts/utils/internet_detect_tool.py
Comment (Ana06) [inline scripts/utils/internet_detect_tool.py:25]: where are the icons stored?

File: scripts/utils/internet_detect_tool.py
Comment (Ana06) [inline scripts/utils/internet_detect_tool.py:26]: @mandiant/commando-vm is there interest in commando VM to use a tool like this? If so, we should keep it generic.

File: scripts/utils/internet_detect_tool.py
Comment (Ana06) [inline scripts/utils/internet_detect_tool.py:19]: I think we should have a hash with TEST_URLs and their EXPECTED_RESPONSE, as if this concrete URL is down, the internet test would fail.

Comment (Ana06) [reply scripts/utils/internet_detect_tool.py:19]: what about testing the adapter in addition?

File: scripts/utils/internet_detect_tool.py
Comment (Ana06) [inline scripts/utils/internet_detect_tool.py:41]: I know this was my idea, but I am thinking now it may be a bad idea:  If internet is enabled and you stop the tool, the internet is still enabled. :thinking:

**Comment by mr-tz** [comment]
How does this script interact with FakeNet? Should FakeNet be handled specifically?

**Comment by Ana06** [comment]
> How does this script interact with FakeNet? Should FakeNet be handled specifically?

If fakenet response is the one we would get with internet, our tool (as other programs like malware) thinks the internet is on. The default fakenet response shouldn't match the ones we are expecting, so without modifying the default fakenet configuration, it will think there is no internet.

@mr-tz do you have a proposal to improve the implementation? What do you exactly mean with handling fakenet specifically?

Note: I think this tool could be hosted in the fakenet repository and even been part of fakenet, it would need to be discussed with the fakenet maintainers to see if it is feasible once we have a working and tested version.

**Comment by mr-tz** [comment]
I'd expect it to not warn users that the Internet is reachable if FakeNet is running.

**Comment by emtuls** [comment]
This raises a good thought.

While the tool works with FakeNet active, you would see a lot of requests generated in order for the tool to function. We can potentially reduce the request interval, but this would cause a delay in detecting internet which is a trade-off we need to consider.

What interval would be fair enough to detect a real internet connection in enough time for a user to be aware while also not being too intrusive in a FakeNet output? Also note, due to adding backup internet connectivity URL's, there are now multiple requests that are issued consecutively, rather than a single one.

File: scripts/utils/internet_detect_tool.py
Comment (emtuls) [reply scripts/utils/internet_detect_tool.py]: Whoops! Thank you for catching that. Added here: https://github.com/mandiant/flare-vm/pull/608

**Comment by emtuls** [comment]
I was able to remove the `PyQt5` dependency. :)

**Comment by emtuls** [comment]
Added nicer tooltips that tell explain what the icons mean: "Internet Detected" vs "Internet Not Detected"
image
image

**Comment by mr-tz** [comment]
I remember some filtering work FakeNet did on Internet connectivity checks to reduce noise. Maybe @tinajohnson can chime in on this?

I'm not sure what a good interval would be, maybe 20 seconds?

**Comment by emtuls** [comment]
Chatting with FakeNet team, it sounds like we can add the tool to "Diverter's ProcessBlackList and HTTPListener's ProcessBlackList": 
see the second paragraph here ("You may also want to enable..." : https://github.com/mandiant/flare-fakenet-ng?tab=readme-ov-file#anonymous-listener

**Comment by emtuls** [comment]
I also updated the tool to use a pink icon when internet is active and gray when not, as well as convert everything over to a VM Package and have it install a scheduled task to run every 2 minutes.

It is a `.pyw` file in order to not have a terminal interface popup for it to run.
image
image

**Comment by emtuls** [comment]
Also note, this will fail to pass the `test_upload` due to requiring certain files to be existent in the main branch already:
https://github.com/mandiant/VM-Packages/pull/1121/files#diff-bb3df9f8d616b7a9b00f92220dafa6cc20883cdd518f04b8e2c3e3993dfe9fe6R15

**Comment by Ana06** [comment]
@emtuls 

> Also note, this will fail to pass the test_upload due to requiring certain files to be existent in the main branch already:
https://github.com/mandiant/VM-Packages/pull/1121/files#diff-bb3df9f8d616b7a9b00f92220dafa6cc20883cdd518f04b8e2c3e3993dfe9fe6R15

We need to split this PR as the package need to succeed to be pushed to MyGet: one PR only for the tool, one PR for the package. Sorry for the confusion.

File: packages/internet_detector.vm/tools/chocolateyinstall.ps1
Comment (Ana06) [inline packages/internet_detector.vm/tools/chocolateyinstall.ps1:13]:
```suggestion
New-Item -Path $toolDir -ItemType Directory -Force -ea 0
```

File: packages/internet_detector.vm/tools/chocolateyinstall.ps1
Comment (Ana06) [inline packages/internet_detector.vm/tools/chocolateyinstall.ps1:16]: use it directly is inside the package

File: packages/internet_detector.vm/tools/chocolateyinstall.ps1
Comment (Ana06) [inline packages/internet_detector.vm/tools/chocolateyinstall.ps1:30]:
```suggestion
# Define the action to be performed in the scheduled task
$action = New-ScheduledTaskAction -Execute $executablePath -Argument "$filePath"
# Define the trigger to run every 2 minutes
$trigger = New-ScheduledTaskTrigger -Once -At (Get-Date) -RepetitionInterval (New-TimeSpan -Minutes 2)
# Create the scheduled task
Register-ScheduledTask -Action $action -Trigger $trigger -TaskName 'Internet Detector' -Force
```

File: packages/internet_detector.vm/tools/chocolateyuninstall.ps1
Comment (Ana06) [inline packages/internet_detector.vm/tools/chocolateyuninstall.ps1:7]: I think we should remove the scheduled task and maybe the toolDir as well

File: packages/internet_detector.vm/tools/internet_detector.pyw
Comment (Ana06) [inline packages/internet_detector.vm/tools/internet_detector.pyw:91]:
```suggestion
    match = re.search(r'<title>(.*?)</title>', data)
```

File: packages/internet_detector.vm/tools/internet_detector.pyw
Comment (Ana06) [inline packages/internet_detector.vm/tools/internet_detector.pyw:7]:
```suggestion
```

File: packages/internet_detector.vm/tools/internet_detector.pyw
Comment (Ana06) [inline packages/internet_detector.vm/tools/internet_detector.pyw:28]: add icons + `background-internet.png`

File: packages/internet_detector.vm/tools/internet_detector.pyw
Comment (Ana06) [inline packages/internet_detector.vm/tools/internet_detector.pyw:35]: [nitpick] run `black -l 120` for consistent styling

File: packages/internet_detector.vm/tools/internet_detector.pyw
Comment (Ana06) [inline packages/internet_detector.vm/tools/internet_detector.pyw:103]: [nitpick] Use the most important Python >3.6 future (f-strings):
```suggestion
                print(f"Internet connectivity detected via URL: {url}")
```

**Review by Ana06 — CHANGES_REQUESTED** [review]

File: packages/internet_detector.vm/tools/chocolateyuninstall.ps1
Comment (emtuls) [reply packages/internet_detector.vm/tools/chocolateyuninstall.ps1:7]: toolDir should be removed via `VM-Uninstall` I believe? https://github.com/mandiant/VM-Packages/blob/921ab89f37a2513fa77571f695415b2160970b03/packages/common.vm/tools/vm.common/vm.common.psm1#L607

**Comment by tinajohnson** [comment]
Apologies for missing the ping here but I did have a conversation about using FakeNet's blacklist config with @emtuls privately. A correction to my original comment on the same - adding the process name to just the Diverter's ProcessBlackList config should do the trick as that will act as a global blacklist.

File: packages/internet_detector.vm/tools/internet_detector.pyw
Comment (Ana06) [inline packages/internet_detector.vm/tools/internet_detector.pyw:151]: I think the transparency effect does not work in a dark taskbar in non-activated Windows (at least I see no difference). What about in addition **setting the taskbar to rose color in the dark mode**? 

I am ok with keeping the transparency if you like it (I will never see it anyway :rofl:).

**Review by Ana06 — CHANGES_REQUESTED** [review]
I have tested it locally and it is very fancy! :tada: Great work @emtuls!

File: packages/internet_detector.vm/tools/chocolateyinstall.ps1
Comment (Ana06) [reply packages/internet_detector.vm/tools/chocolateyinstall.ps1:26]: after doing some more testing, it seems to work without admin rights. :confused:

Comment (Ana06) [reply packages/internet_detector.vm/tools/chocolateyinstall.ps1:26]: We should keep it consistent with how we run the scheduled task, is it running with admin rights?

File: packages/internet_detector.vm/tools/internet_detector.pyw
Comment (Ana06) [inline packages/internet_detector.vm/tools/internet_detector.pyw:295]: I would move this `# Rose Color = #ff5200c3` comment to where the `ROSE_ACCENT_PALETTE` constant is defined

**Review by Ana06 — COMMENTED** [review]
Great work @emtuls :tada: 

image

Just some minor improvement suggestions :wink: 
- can we change the pink icon to have more contrast with the pink taskbask? (for example with a more intense/lighter pink) it is a bit difficult to see when it is hidden and it does not hurt when it is not:
image
image

- it would be nice that the icon is always visible. It seems I can move it manually to the taskbar and then it stays there. It would be nice if the tool could place it there. I propose you open an issue your this and research if it is possible after this PR has been merged. As it can be placed manually in the taskbar, I do not consider this a blocker and I would like to start using the tool! :tada:

**Comment by emtuls** [comment]
> Great work @emtuls 🎉
> 
> image
> 
> Just some minor improvement suggestions 😉
> 
> * can we change the pink icon to have more contrast with the pink taskbask? (for example with a more intense/lighter pink) it is a bit difficult to see when it is hidden and it does not hurt when it is not:
>   image
>   image
> * it would be nice that the icon is always visible. It seems I can move it manually to the taskbar and then it stays there. It would be nice if the tool could place it there. I propose you open an issue your this and research if it is possible after this PR has been merged. As it can be placed manually in the taskbar, I do not consider this a blocker and I would like to start using the tool! 🎉

I have updated it to just use the colored version, which I think looks great with the pink taskbar.

Unfortunately, I don't think there's a way to force the icon to stay on the tray in the taskbar...For me, it starts there, but sometimes gets moved, but it can be moved back down manually. Couldn't find any way to force it to stay on the bottom.

File: packages/internet_detector.vm/tools/default.ini
Comment (Ana06) [inline packages/internet_detector.vm/tools/default.ini:264]: we can not exclude `pythonw.exe`. It is too generic and it could be needed to run malware.

Two options:
### duplicate `pythonw.exe` with a different name [ugly temporary workaround]
In the package, duplicate `pythonw.exe` in the same directory with name `pythonw_internet_detector.exe`. Use `pythonw_internet_detector.exe` to run the tool and in the fakenet log. It is a bit ugly, but we could use it as a temporary workaround to be able to merge this PR and start testing this tool.

### build the tool
We could build the tool into an executable in the package. I see [capa doing just: `python -m build`](https://github.com/mandiant/capa/blob/master/.github/workflows/publish.yml), so it may be also not difficult to do. @mr-tz and @williballenthin may be able to help with how to best implement it. :pray:

File: packages/internet_detector.vm/tools/chocolateyinstall.ps1
Comment (Ana06) [inline packages/internet_detector.vm/tools/chocolateyinstall.ps1:32]:
```suggestion
# Move updated `default.ini` with tool added to ProcessBlackList to FakeNet Config Directory
# TODO: Contribute this change to the fakenet default configuration
```

File: packages/internet_detector.vm/tools/chocolateyinstall.ps1
Comment (Ana06) [inline packages/internet_detector.vm/tools/chocolateyinstall.ps1:36]: I think you are doing something wrong moving the file, as the file I see in my fakenet default configuration after installing this package does not have your changes. Please test this out @emtuls!

Also, I would copy instead of moving, so that we have a copy of the file in case it is needed it.

**Review by Ana06 — COMMENTED** [review]
tool looks good! we need to figure out

File: packages/internet_detector.vm/tools/chocolateyinstall.ps1
Comment (Ana06) [inline packages/internet_detector.vm/tools/chocolateyinstall.ps1:5]: Wondering if `Networking` is a better category than `Utilities`. @mandiant/flare-vm opinions?

**Review by Ana06 — CHANGES_REQUESTED** [review]
If I have internet in the VM and I start fakenet, I expect the tool to keep showing me there is internet. But this is not the case at the moment (both with and without the config file). Is this the fakenet bug reported in https://github.com/mandiant/flare-fakenet-ng/issues/190?

File: packages/internet_detector.vm/tools/chocolateyinstall.ps1
Comment (Ana06) [inline packages/internet_detector.vm/tools/chocolateyinstall.ps1:36]: **[IMPORTANT]**As mentioned in a previous comment, this is not working. You are copying the config file to `C:\Tools\fakenet\fakenet3.2-alpha` instead of `C:\Tools\fakenet\fakenet3.2-alpha\configs`, so the config is not being applied.

I also think we should use the tool directory directly instead of the shortcut, as if the user removes the shortcut in the desktop everything should continue working.

```suggestion
$fakenetConfigDir = Get-ChildItem "C:\Tools\fakenet\*\configs"
if ($fakenetConfigDir) {
    Copy-Item "$packageToolDir\default.ini" -Destination $fakenetConfigDir
```

**Review by Ana06 — CHANGES_REQUESTED** [review]

File: packages/internet_detector.vm/tools/chocolateyinstall.ps1
Comment (Ana06) [reply packages/internet_detector.vm/tools/chocolateyinstall.ps1:32]: Alternatively (better more robust solution): Move this to our fakenet package in this PR.

We can still open the PR/issue in fakenet afterwards to discuss if porting the change upstream is an option, as otherwise we will have to keep this file in sync in our package which is not ideal.

File: packages/fakenet-ng.vm/tools/chocolateyinstall.ps1
Comment (Ana06) [inline packages/fakenet-ng.vm/tools/chocolateyinstall.ps1:39]: [nitpick]
```suggestion
  # IMPORTANT: Keep our modified `default.ini` in-sync on updates to package.
```

There is nothing TODO if no update is done :wink:

**Review by Ana06 — APPROVED** [review]
Thanks for all the work @emtuls, two minor improvement suggestions, but I think we are good to merge this! :wink:

**Review by Ana06 — APPROVED** [review]
We are ready to finally get this in and get my pink dynamic analysis VM! :pink_heart:

**Pull request merged**
